ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -8379,7 +8402,9 @@ private ColsAndTypes deriveFileSinkColTypes(
       String colName = colInfo.getInternalName();  //default column name
       if (columns != null) {
         FieldSchema col = new FieldSchema();
-        if (!("".equals(nm[0])) && nm[1] != null) {
+        if (i < withColList.size()) {
```

Comment:
Using a random name when <with column list> is shorter than the input. This follows the behavior of non-materialized CTEs.
If I understand correctly, with-column-list is supported only when the number of columns is exactly equal to that of the query expression. So, it could be an option to make a query fail if `colInfos.size() != withColList.size()`.